massa-protocol-worker/src/handlers/peer_handler/models.rs

```diff
@@ -16,15 +16,114 @@ const THREE_DAYS_MS: u64 = 3 * 24 * 60 * 60 * 1_000_000;
 
 pub type InitialPeers = HashMap<PeerId, HashMap<SocketAddr, TransportType>>;
 
+#[derive(Clone, Eq, PartialEq)]
+pub struct ConnectionMetadata {
+    pub last_try: Option<MassaTime>,
+    pub last_success: Option<MassaTime>,
+    pub last_failure: Option<MassaTime>,
+    random_priority: u64,
+}
+
+impl Default for ConnectionMetadata {
+    fn default() -> Self {
+        ConnectionMetadata {
+            last_try: Default::default(),
+            last_success: Default::default(),
+            last_failure: Default::default(),
+            random_priority: thread_rng().gen(),
+        }
+    }
+}
+
+impl Ord for ConnectionMetadata {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.partial_cmp(other).unwrap()
+    }
+}
+
+// Priorisation of a peer compared to another one
+// Greater = More Prio        Lesser = Less prio
+impl PartialOrd for ConnectionMetadata {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        let failure_check = match (self.last_failure, other.last_failure) {
+            (Some(sf), Some(of)) => Some(sf.cmp(&of)),
+            (Some(_), None) => Some(Ordering::Greater),
+            (None, Some(_)) => Some(Ordering::Less),
+            (None, None) => None,
+        };
+        if let Some(res) = failure_check {
+            return Some(res);
+        }
+        let success_check = match (self.last_success, other.last_success) {
+            (Some(ss), Some(os)) => Some(ss.cmp(&os).reverse()),
+            (Some(_), None) => Some(Ordering::Less),
+            (None, Some(_)) => Some(Ordering::Greater),
+            (None, None) => None,
+        };
+        if let Some(res) = success_check {
+            return Some(res);
+        }
+        let try_check = match (self.last_try, other.last_try) {
+            (Some(st), Some(ot)) => Some(st.cmp(&ot).reverse()),
+            (Some(_), None) => Some(Ordering::Less),
+            (None, Some(_)) => Some(Ordering::Greater),
+            (None, None) => None,
+        };
+        if let Some(res) = try_check {
+            Some(res)
+        } else {
+            Some(self.random_priority.cmp(&other.random_priority))
+        }
+    }
+}
+
+impl ConnectionMetadata {
+    // Only used in tests
+    #[allow(dead_code)]
+    pub fn edit(self, data_type: usize, data: Option<MassaTime>) -> ConnectionMetadata {
+        match data_type {
+            0 => ConnectionMetadata {
+                last_failure: data,
+                ..self
+            },
+            1 => ConnectionMetadata {
+                last_success: data,
+                ..self
+            },
+            2 => ConnectionMetadata {
+                last_try: data,
+                ..self
+            },
+            _ => unreachable!("connection metadata data_type not recognized: {data_type}"),
+        }
+    }
+    pub fn new_failure(self) -> ConnectionMetadata {
```

Comment:
you can take a mutable ref and update self instead of return new ConnectionMetadata